File: lib/domain/agricultural_task/gateways/agricultural_task_gateway.rb
# frozen_string_literal: true

module Domain
  module AgriculturalTask
    module Gateways
      class AgriculturalTaskGateway
        def list_user_owned_tasks(user_id:, query: nil)
          raise NotImplementedError, "Subclasses must implement list_user_owned_tasks"
        end

        def list_reference_tasks(query: nil)
          raise NotImplementedError, "Subclasses must implement list_reference_tasks"
        end

        def list_user_and_reference_tasks(user_id:, query: nil)
          raise NotImplementedError, "Subclasses must implement list_user_and_reference_tasks"
        end

        # 詳細 DTO。認可は Interactor 側（R0）。
        def find_agricultural_task_show_detail(id)
          raise NotImplementedError, "Subclasses must implement find_agricultural_task_show_detail"
        end

        def find_agricultural_task_loaded_bundle!(id, for_edit:)
          raise NotImplementedError, "Subclasses must implement find_agricultural_task_loaded_bundle!"
        end

        def find_by_id(id)
          raise NotImplementedError, "Subclasses must implement find_by_id"
        end

        def create_for_user(user, attrs)
          raise NotImplementedError, "Subclasses must implement create_for_user"
        end

        # @param selected_crop_ids [Array<Integer>, nil] nil のとき作物テンプレート同期を行わない（API 等）
        def update_for_user(user, id, attrs, selected_crop_ids: nil)
          raise NotImplementedError, "Subclasses must implement update_for_user"
        end

        def soft_delete_with_undo(user:, task_id:, auto_hide_after:, translator:)
          raise NotImplementedError, "Subclasses must implement soft_delete_with_undo"
        end

        # 作物選択 UI: 当該農業作業に紐づく CropTaskTemplate の作物 ID 一覧
        def linked_crop_ids_for_task_templates(agricultural_task_id)
          raise NotImplementedError, "Subclasses must implement linked_crop_ids_for_task_templates"
        end

        # HTML 新規フォーム用の未保存タスク（永続化しない）
        def build_blank_agricultural_task_for_master_form(user)
          raise NotImplementedError, "Subclasses must implement build_blank_agricultural_task_for_master_form"
        end

        # マスタCRUD 作成失敗時: 送信属性で検証エラー状態を再現（永続化しない）
        def build_after_create_failure_agricultural_task_for_master_form!(user:, attributes:)
          raise NotImplementedError, "Subclasses must implement build_after_create_failure_agricultural_task_for_master_form!"
        end

        # HTML 編集の作物選択データ読込: update 時は送信パラメータを反映したプレビュー用エンティティを返す（永続化しない）
        # @param base_entity [Domain::AgriculturalTask::Entities::AgriculturalTaskEntity]
        # @return [Domain::AgriculturalTask::Entities::AgriculturalTaskEntity]
        def preview_agricultural_task_for_edit_crop_selection(base_entity:, user:, agricultural_task_params:)
          raise NotImplementedError, "Subclasses must implement preview_agricultural_task_for_edit_crop_selection"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/agricultural_task/gateways/agricultural_task_gateway.rs
use crate::agricultural_task::dtos::AgriculturalTaskShowDetail;
use crate::agricultural_task::entities::AgriculturalTaskEntity;
use crate::shared::attr::AttrMap;
use crate::shared::dtos::Error;
use crate::shared::user::User;

/// Ruby: `Domain::AgriculturalTask::Gateways::AgriculturalTaskGateway`
pub trait AgriculturalTaskGateway: Send + Sync {
    fn list_user_owned_tasks(
        &self,
        user_id: i64,
        query: Option<&str>,
    ) -> Result<Vec<AgriculturalTaskEntity>, Box<dyn std::error::Error + Send + Sync>>;

    fn list_reference_tasks(
        &self,
        query: Option<&str>,
    ) -> Result<Vec<AgriculturalTaskEntity>, Box<dyn std::error::Error + Send + Sync>>;

    fn list_user_and_reference_tasks(
        &self,
        user_id: i64,
        query: Option<&str>,
    ) -> Result<Vec<AgriculturalTaskEntity>, Box<dyn std::error::Error + Send + Sync>>;

    fn find_agricultural_task_show_detail(
        &self,
        id: i64,
    ) -> Result<AgriculturalTaskShowDetail, Box<dyn std::error::Error + Send + Sync>>;

    fn find_by_id(
        &self,
        id: i64,
    ) -> Result<AgriculturalTaskEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn find_by_reference_and_name(
        &self,
        name: &str,
    ) -> Result<Option<AgriculturalTaskEntity>, Box<dyn std::error::Error + Send + Sync>>;

    fn find_by_user_id_and_name(
        &self,
        user_id: i64,
        name: &str,
    ) -> Result<Option<AgriculturalTaskEntity>, Box<dyn std::error::Error + Send + Sync>>;

    fn create(
        &self,
        attrs: AttrMap,
    ) -> Result<AgriculturalTaskEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn update(
        &self,
        id: i64,
        attrs: AttrMap,
    ) -> Result<AgriculturalTaskEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn within_transaction<F, T>(&self, block: F) -> T
    where
        F: FnOnce() -> T;

    fn soft_delete_with_undo(
        &self,
        user: &User,
        task_id: i64,
        auto_hide_after: i64,
        toast_message: &str,
    ) -> Result<SoftDeleteUndoResult, Box<dyn std::error::Error + Send + Sync>>;
}

#[derive(Debug, Clone)]
pub enum SoftDeleteUndoResult {
    Success { undo: crate::agricultural_task::dtos::UndoEntity },
    Failure { error: Error },
}
